# P01_02_Data_Scraping

In [5]:
import os, sys
import requests
from urllib.parse import urljoin, urlparse
from bs4 import BeautifulSoup as bs
import re
import pandas as pd


## Read Data

the page html file downloaded in Data directory and it is loaded below.

In [47]:
file = open("../Data/ICME_WebScr_data.html", "r")
html = file.read()
# print(html)

In [15]:
soup = bs(html)
# tds = soup.find_all('tr')

In [23]:
column_names_1 = soup.find_all('tr')
print(column_names_1[0])

<tr align="center"><td><b>Disturbance Y/M/D (UT) <a href="#(a)">(a)</a></b> </td><td colspan="2">
<b>ICME Plasma/Field Start, End Y/M/D (UT) <a href="#(b)">(b)</a> </b> </td><td colspan="2">
<b>Comp. Start, End (Hrs wrt. Plasma/ Field) <a href="#(c)">(c)</a></b> </td><td colspan="2">
<b>MC Start, End (Hrs wrt. Plasma/ Field) <a href="#(d)">(d)</a></b> </td><td>
<b>BDE? <a href="#(e)">(e)</a></b> </td><td>
<b>BIF? <a href="#(f)">(f)</a></b> </td><td>
<b> Qual. <a href="#(g)">(g)</a></b></td><td>
<b>dV (km/s) <a href="#(h)">(h)</a></b> </td><td>
<b>V_ICME (km/s) <a href="#(i)">(i)</a></b> </td><td>
<b>V_max (km/s) <a href="#(j)">(j)</a></b> </td><td>
<b>B (nT) <a href="#(k)">(k)</a></b> </td><td>
<b>MC? <a href="#(l)">(l)</a></b> </td><td>
<b>Dst (nT) <a href="#(m)">(m)</a></b> </td><td>
<b>V_transit (km/s) <a href="#(n)">(n)</a></b> </td><td>
<b>LASCO CME Y/M/D (UT)<a href="#(o)"> (o)</a></b></td></tr>


In [24]:
len(column_names_1)

562

In [25]:
column_name_2  = column_names_1[0].find_all('b')


In [26]:
column_name_2[0].text.strip()


'Disturbance Y/M/D (UT) (a)'

In [27]:
len(column_name_2)


15

## Column names

In [28]:
Data_column_names = []
for col in column_name_2:
#     print(col.text.strip())
    Data_column_names.append(col.text.strip())
Data_column_names

['Disturbance Y/M/D (UT) (a)',
 'ICME Plasma/Field Start, End Y/M/D (UT) (b)',
 'Comp. Start, End (Hrs wrt. Plasma/ Field) (c)',
 'MC Start, End (Hrs wrt. Plasma/ Field) (d)',
 'BDE? (e)',
 'BIF? (f)',
 'Qual. (g)',
 'dV (km/s) (h)',
 'V_ICME (km/s) (i)',
 'V_max (km/s) (j)',
 'B (nT) (k)',
 'MC? (l)',
 'Dst (nT) (m)',
 'V_transit (km/s) (n)',
 'LASCO CME Y/M/D (UT) (o)']

In [29]:
len(Data_column_names)


15

In [48]:
data_1 = soup.find_all('tr', align="center")
data_1[0]

<tr align="center"><td><b>Disturbance Y/M/D (UT) <a href="#(a)">(a)</a></b> </td><td colspan="2">
<b>ICME Plasma/Field Start, End Y/M/D (UT) <a href="#(b)">(b)</a> </b> </td><td colspan="2">
<b>Comp. Start, End (Hrs wrt. Plasma/ Field) <a href="#(c)">(c)</a></b> </td><td colspan="2">
<b>MC Start, End (Hrs wrt. Plasma/ Field) <a href="#(d)">(d)</a></b> </td><td>
<b>BDE? <a href="#(e)">(e)</a></b> </td><td>
<b>BIF? <a href="#(f)">(f)</a></b> </td><td>
<b> Qual. <a href="#(g)">(g)</a></b></td><td>
<b>dV (km/s) <a href="#(h)">(h)</a></b> </td><td>
<b>V_ICME (km/s) <a href="#(i)">(i)</a></b> </td><td>
<b>V_max (km/s) <a href="#(j)">(j)</a></b> </td><td>
<b>B (nT) <a href="#(k)">(k)</a></b> </td><td>
<b>MC? <a href="#(l)">(l)</a></b> </td><td>
<b>Dst (nT) <a href="#(m)">(m)</a></b> </td><td>
<b>V_transit (km/s) <a href="#(n)">(n)</a></b> </td><td>
<b>LASCO CME Y/M/D (UT)<a href="#(o)"> (o)</a></b></td></tr>

In [32]:
len(data_1)


562

the first element is the headers or column names. check out the number of columns


In [33]:
data_1[0].text.strip().split('\n')


['Disturbance Y/M/D (UT) (a) ',
 'ICME Plasma/Field Start, End Y/M/D (UT) (b)  ',
 'Comp. Start, End (Hrs wrt. Plasma/ Field) (c) ',
 'MC Start, End (Hrs wrt. Plasma/ Field) (d) ',
 'BDE? (e) ',
 'BIF? (f) ',
 ' Qual. (g)',
 'dV (km/s) (h) ',
 'V_ICME (km/s) (i) ',
 'V_max (km/s) (j) ',
 'B (nT) (k) ',
 'MC? (l) ',
 'Dst (nT) (m) ',
 'V_transit (km/s) (n) ',
 'LASCO CME Y/M/D (UT) (o)']

In [34]:
a = list(data_1[0])
len(a)

15

There are 15 columns in table. The next element has 18 elements which are the data we looking for. let's check out the number of elements.



In [35]:
a = list(data_1[1])
len(a)

18

make a list of data and iterate through and extract the data.

In [36]:
data_ext = []
for i in range(0,18):
    data_ext.append(a[i].text.strip())
#     print(a[i].text.strip())

In [37]:
# data_ext

we have to iterate through all 562 elements of data. 

note that some are just header with 15 sub elements, and some are the data with 18. let's gather all and later filter out data.

In [38]:
num_ite_1 = len(data_1)

# keep data in data_ext list
data_ext = []
for i in range(0, num_ite_1):
    
    # make a list
    a = list(data_1[i])
    num_ite_2 = len(a)
#     print(num_ite_2)
    
    collector = []
    for j in range(0, num_ite_2):
        try:
            collector.append(a[j].text.strip())
        except:
            pass
    data_ext.append(collector)


In [39]:
len(data_ext[1])

18

In [40]:
real_data = []
for i in range(len(data_ext)):
    if len(data_ext[i]) == 18:
        real_data.append(data_ext[i])
        

In [41]:
len(real_data)

513

In [51]:
Data_column_names

['Disturbance Y/M/D (UT) (a)',
 'ICME Plasma/Field Start, End Y/M/D (UT) (b)',
 'Comp. Start, End (Hrs wrt. Plasma/ Field) (c)',
 'MC Start, End (Hrs wrt. Plasma/ Field) (d)',
 'BDE? (e)',
 'BIF? (f)',
 'Qual. (g)',
 'dV (km/s) (h)',
 'V_ICME (km/s) (i)',
 'V_max (km/s) (j)',
 'B (nT) (k)',
 'MC? (l)',
 'Dst (nT) (m)',
 'V_transit (km/s) (n)',
 'LASCO CME Y/M/D (UT) (o)']

Check out the column names and edit the columns list and pass it to data frame.

<p align="center">
  <img width="1000" src="../Assets/column_names.PNG" >
</p>

In [58]:
Data_column_names = ['Disturbance Y/M/D (UT) (a)',
                     
                     'ICME Plasma/Field Start Y/M/D (UT) (b)',
                     'ICME Plasma/Field End Y/M/D (UT) (b)',
                     
                     'Comp. Start (Hrs wrt. Plasma/ Field) (c)',
                     'Comp. End (Hrs wrt. Plasma/ Field) (c)',
                     
                     'MC Start (Hrs wrt. Plasma/ Field) (d)',
                     'End (Hrs wrt. Plasma/ Field) (d)',
                     
                     'BDE? (e)',
                     'BIF? (f)',
                     'Qual. (g)',
                     'dV (km/s) (h)',
                     'V_ICME (km/s) (i)',
                     'V_max (km/s) (j)',
                     'B (nT) (k)',
                     'MC? (l)',
                     'Dst (nT) (m)',
                     'V_transit (km/s) (n)',
                     'LASCO CME Y/M/D (UT) (o)']

In [59]:
real_data_pd = pd.DataFrame(real_data, columns=Data_column_names)
real_data_pd.head()

Disturbance Y/M/D (UT) (a) ICME Plasma/Field Start Y/M/D (UT) (b)  \
0            1996/05/27 1500                        1996/05/27 1500   
1            1996/07/01 1320                        1996/07/01 1800   
2            1996/08/07 0600                        1996/08/07 1200   
3            1996/12/23 1600                        1996/12/23 1700   
4            1997/01/10 0104                        1997/01/10 0400   

  ICME Plasma/Field End Y/M/D (UT) (b)  \
0                      1996/05/29 0300   
1                      1996/07/02 1100   
2                      1996/08/08 1000   
3                      1996/12/25 1100   
4                      1997/01/11 0200   

  Comp. Start (Hrs wrt. Plasma/ Field) (c)  \
0                                      ...   
1                                      ...   
2                                      ...   
3                                      ...   
4                                      ...   

  Comp. End (Hrs wrt. Plasma/ Field) (c)  \
0                                    ...   
1                                    ...   
2                                    ...   
3                                    ...   
4                                    ...   

  MC Start (Hrs wrt. Plasma/ Field) (d) End (Hrs wrt. Plasma/ Field) (d)  \
0                                     0                               +4   
1                                     0                                0   
2                                     0                                0   
3                                   +10                                0   
4                                     0                                0   

  BDE? (e) BIF? (f) Qual. (g) dV (km/s) (h) V_ICME (km/s) (i)  \
0        N      ...         2             0               370   
1        N      ...         3            40               360   
2        N      ...         2            10               350   
3        N      ...         2            20               360   
4        Y      ...         1         100 S               450   

  V_max (km/s) (j) B (nT) (k) MC? (l) Dst (nT) (m) V_transit (km/s) (n)  \
0              400          9       2          -33                  ...   
1              370         11       2          -20                  ...   
2              380          7       2          -23                  ...   
3              420         10       2          -18                  435   
4              460         14       2          -78                  507   

  LASCO CME Y/M/D (UT) (o)  
0                           
1                           
2                           
3        1996/12/19 1630 H  
4        1997/01/06 1510 H

In [60]:
real_data_pd.tail()

Disturbance Y/M/D (UT) (a) ICME Plasma/Field Start Y/M/D (UT) (b)  \
508            2020/10/05 0806                        2020/10/05 1700   
509            2020/12/10 0210                        2020/12/10 2300   
510            2021/02/23 0200                        2021/02/23 0800   
511            2021/05/12 0640                        2021/05/12 1500   
512            2021/05/26 1246                        2021/05/26 2200   

    ICME Plasma/Field End Y/M/D (UT) (b)  \
508                      2020/10/06 1400   
509                      2020/12/13 0400   
510                      2021/02/24 1500   
511                      2021/05/14 2000   
512                      2021/05/28 0300   

    Comp. Start (Hrs wrt. Plasma/ Field) (c)  \
508                                      ...   
509                                      ...   
510                                      ...   
511                                      ...   
512                                      ...   

    Comp. End (Hrs wrt. Plasma/ Field) (c)  \
508                                    ...   
509                                    ...   
510                                    ...   
511                                    ...   
512                                    ...   

    MC Start (Hrs wrt. Plasma/ Field) (d) End (Hrs wrt. Plasma/ Field) (d)  \
508                                   ...                              ...   
509                                   ...                              ...   
510                                   ...                              ...   
511                                   ...                              ...   
512                                   ...                              ...   

    BDE? (e) BIF? (f) Qual. (g) dV (km/s) (h) V_ICME (km/s) (i)  \
508        Y      ...         2            20               350   
509        Y      ...         2           110               440   
510        Y      ...        3W            40               510   
511        Y      ...         2           120               410   
512      ...      ...         3            40               410   

    V_max (km/s) (j) B (nT) (k) MC? (l) Dst (nT) (m) V_transit (km/s) (n)  \
508              370         10       2        -40 Q                  300   
509              570          5       0        -10 Q                  720   
510              590          5       0        -31 Q                  670   
511              500          8       2        -61 Q                  620   
512              430          9       0        -28 Q              550-700   

                LASCO CME Y/M/D (UT) (o)  
508                   2020/09/29 1209 D?  
509                   2020/12/07 1624H D  
510                    2021/02/20 1136 D  
511                    2021/05/09 1123 D  
512  2021/05/23 0953, 1138, 1738, 2525 D

In [61]:
real_data_pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 513 entries, 0 to 512
Data columns (total 18 columns):
Disturbance Y/M/D (UT) (a)                  513 non-null object
ICME Plasma/Field Start Y/M/D (UT) (b)      513 non-null object
ICME Plasma/Field End Y/M/D (UT) (b)        513 non-null object
Comp. Start (Hrs wrt. Plasma/ Field) (c)    513 non-null object
Comp. End (Hrs wrt. Plasma/ Field) (c)      513 non-null object
MC Start (Hrs wrt. Plasma/ Field) (d)       513 non-null object
End (Hrs wrt. Plasma/ Field) (d)            513 non-null object
BDE? (e)                                    513 non-null object
BIF? (f)                                    513 non-null object
Qual. (g)                                   513 non-null object
dV (km/s) (h)                               513 non-null object
V_ICME (km/s) (i)                           513 non-null object
V_max (km/s) (j)                            513 non-null object
B (nT) (k)                                  513 non-null object

In [62]:
real_data_pd.describe().T

count unique              top freq
Disturbance Y/M/D (UT) (a)                 513    513  2001/09/30 1924    1
ICME Plasma/Field Start Y/M/D (UT) (b)     513    513  2000/07/14 1700    1
ICME Plasma/Field End Y/M/D (UT) (b)       513    513  2012/07/23 1600    1
Comp. Start (Hrs wrt. Plasma/ Field) (c)   513     31              ...  236
Comp. End (Hrs wrt. Plasma/ Field) (c)     513     41              ...  236
MC Start (Hrs wrt. Plasma/ Field) (d)      513     19              ...  410
End (Hrs wrt. Plasma/ Field) (d)           513     41              ...  410
BDE? (e)                                   513      4                Y  346
BIF? (f)                                   513      4              ...  400
Qual. (g)                                  513      5                2  255
dV (km/s) (h)                              513     63                0   62
V_ICME (km/s) (i)                          513     51              400   36
V_max (km/s) (j)                           513     68              460   30
B (nT) (k)                                 513     25                8   60
MC? (l)                                    513      4                2  178
Dst (nT) (m)                               513    223              -36    9
V_transit (km/s) (n)                       513    126              ...  247
LASCO CME Y/M/D (UT) (o)                   513    252                   235

## Save data

In [63]:
file_path = "../Data/"
file_name = "Data_ICME_scraped.csv"
real_data_pd.to_csv(file_path+file_name)

In [377]:
# csv_path = "../Data/Data_ICME_scraped.csv"
# df = pd.read_csv(csv_path, index_col='0', parse_dates=True)